# **Stemming Tweet Classification**

1.   Stanley Pudjowibowo 20.K1.0006
2.   Daniel Hartanto 20.K1.0008
3.   Jonathan Kevin Giustino 20.K1.0009



In [ ]:
!git clone https://github.com/rizalespe/Dataset-Sentimen-Analisis-Bahasa-Indonesia.git
!pip install PySastrawi

Cloning into 'Dataset-Sentimen-Analisis-Bahasa-Indonesia'...
remote: Enumerating objects: 169, done.
remote: Counting objects: 100% (46/46), done.
remote: Compressing objects: 100% (46/46), done.
remote: Total 169 (delta 24), reused 0 (delta 0), pack-reused 123
Receiving objects: 100% (169/169), 164.89 KiB | 3.30 MiB/s, done.
Resolving deltas: 100% (80/80), done.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 210.6/210.6 kB 6.4 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
import nltk
import re
from nltk.corpus import stopwords
from nltk.tokenize import TweetTokenizer
from sklearn.model_selection import train_test_split
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
np.random.seed(0)
# membuat stemmer
factory = StemmerFactory()
stemmer_sastrawi = factory.create_stemmer()
# nltk.download('stopwords')
# nltk.download('averaged_perceptron_tagger')
# nltk.download('maxent_ne_chunker')
# nltk.download('words')
# stopwords_list = stopwords.words('english')

# tokenizer
tk = TweetTokenizer()

np.random.seed(0)
# read data csv
corpus = pd.read_csv('/content/Dataset-Sentimen-Analisis-Bahasa-Indonesia/dataset_tweet_sentiment_pilkada_DKI_2017.csv')
# read stopword csv
stopword_list = pd.read_csv('/content/Dataset-Sentimen-Analisis-Bahasa-Indonesia/stopword_tweet_pilkada_DKI_2017.csv')
# read emoji csv
emojis = pd.read_csv('/content/Dataset-Sentimen-Analisis-Bahasa-Indonesia/master_emoji.csv')

In [ ]:
# changing <Angry Face> into <Angry_Face>
def space(text):
  return re.sub("\s", "_", text)
emojis['emot']=emojis['Special Tag'].apply(lambda x: space(x))
emojis

ID       Emoji Sentiment  \
0      1           😛  negative   
1      2           😠  negative   
2      3           💣  negative   
3      4           💔  negative   
4      5           😕  negative   
..   ...         ...       ...   
160  161           💡  positive   
161  162           📈  positive   
162  163           💯  positive   
163  164           🆗  positive   
164  165  👨🏽‍❤️‍💋‍👩🏼  positive   

                                           Makna Emoji  \
0                           Face With Stuck-Out Tongue   
1                                           Angry Face   
2                                                 Bomb   
3                                         Broken Heart   
4                                        Confused Face   
..                                                 ...   
160                                         Light Bulb   
161                                   Chart Increasing   
162                                     Hundred Points   
163                                          OK Button   
164  Kiss - Man Medium Skin Tone Woman MediumLight ...   

                                           Special Tag  \
0                        <FACE WITH STUCK-OUT TOUNGES>   
1                                         <Angry Face>   
2                                               <Bomb>   
3                                       <Broken Heart>   
4                                      <Confused Face>   
..                                                 ...   
160                                       <Light Bulb>   
161                                 <Chart Increasing>   
162                                   <Hundred Points>   
163                                        <OK Button>   
164  <Kiss - Man Medium Skin Tone Woman MediumLight...   

                                                  emot  
0                       _<FACE_WITH_STUCK-OUT_TOUNGES>  
1                                         <Angry_Face>  
2                                               <Bomb>  
3                                       <Broken_Heart>  
4                                      <Confused_Face>  
..                                                 ...  
160                                       <Light_Bulb>  
161                                 <Chart_Increasing>  
162                                   <Hundred_Points>  
163                                        <OK_Button>  
164  <Kiss_-_Man_Medium_Skin_Tone_Woman_MediumLight...  

[165 rows x 6 columns]

In [ ]:
corpus

Id Sentiment Pasangan Calon  \
0      1  negative     Agus-Sylvi   
1      2  negative     Agus-Sylvi   
2      3  negative     Agus-Sylvi   
3      4  negative     Agus-Sylvi   
4      5  negative     Agus-Sylvi   
..   ...       ...            ...   
895  896  positive    Anies-Sandi   
896  897  positive    Anies-Sandi   
897  898  positive    Anies-Sandi   
898  899  positive    Anies-Sandi   
899  900  positive    Anies-Sandi   

                                            Text Tweet  
0    Banyak akun kloning seolah2 pendukung #agussil...  
1    #agussilvy bicara apa kasihan yaa...lap itu ai...  
2    Kalau aku sih gak nunggu hasil akhir QC tp lag...  
3    Kasian oh kasian dengan peluru 1milyar untuk t...  
4    Maaf ya pendukung #AgusSilvy..hayo dukung #Ani...  
..                                                 ...  
895  Kali saja bpk @aniesbaswedan @sandiuno lihat, ...  
896  Kita harus dapat merangkul semua orang tanpa b...  
897  Ini jagoanku dibidang digital <Smiling Face Wi...  
898               #PesanBijak #OkeOce #GubernurGu3 ...  
899  Sandiaga: Bangun Rumah DP 0% Lebih Simpel Diba...  

[900 rows x 4 columns]

In [ ]:
#change sentiment into index
idx2category = corpus['Sentiment'].unique()
category2idx = dict(zip(idx2category, np.arange(len(idx2category))))

print("idx2category", idx2category)
print("category2idx", category2idx)
corpus['Sentiment'] = corpus['Sentiment'].apply(lambda x : category2idx[x])

idx2category ['negative' 'positive']
category2idx {'negative': 0, 'positive': 1}


In [ ]:
train, test = train_test_split(corpus, test_size=0.3)
train

Id  Sentiment Pasangan Calon  \
76    77          0     Agus-Sylvi   
48    49          0     Agus-Sylvi   
664  665          0    Anies-Sandi   
263  264          1     Agus-Sylvi   
727  728          0    Anies-Sandi   
..   ...        ...            ...   
835  836          1    Anies-Sandi   
192  193          1     Agus-Sylvi   
629  630          0    Anies-Sandi   
559  560          1    Ahok-Djarot   
684  685          0    Anies-Sandi   

                                            Text Tweet  
76   Sah-sah aja, Justru itu menjadi tolak ukur ked...  
48   Masih mau mencaci maki #AHY dan Pepo Memo nya?...  
664  DP 0 persen ternyata beda dengan DP 0 rupiah, ...  
263                   #AHY dukung #HariSantriNasional   
727  Hati2 terprovokasi ya teman. Ini bukan wajah i...  
..                                                 ...  
835  Siiiip pak. Selalu bantu dan kawal #AniesSandi...  
192  Salut! @AgusYudhoyono & Mpok Silvy dengan meng...  
629  Lho, Lagu Anies-Sandi Menjiplak Nyanyian Yahud...  
559  Ayo coblos yang bijaksana untuk JAKARTA yang l...  
684  Ente Kafirun nantangin laskar #AniesSandi ??! ...  

[630 rows x 4 columns]

In [ ]:
# Preprocessing with stemmers
def preprocessing_sastrawi(text):
    token = None
    # removing url
    text= re.sub(r'http\S+', '', text)
    # removing special character ( username and hashtag)
    text = re.sub(r"@\w+", '@username', text) # mengganti username
    text = re.sub(r"#\w+", '#hastag', text) # mengganti hastag
    # changing emoticon
    for i in range(len(emojis)):
      text = re.sub(emojis['Special Tag'][i], emojis['emot'][i], text)
    # Punctuation removal (rmoving symbols)
    text=re.sub(r"[^a-zA-Z0-9\s<>]", "", text)
    # Case folding
    text=text.lower()
    # Tokenizer
    token_text= tk.tokenize(text)
    # Stopword Removal and Stemming
    token = [stemmer_sastrawi.stem(word) for word in token_text if word not in stopword_list]
    return " ".join(token)
# Preprocessing without stemmers
def preprocessing_without_stemmer(text):
    token = None
    # removing url
    text= re.sub(r'http\S+', '', text)
    # removing special character ( username and hashtag)
    text = re.sub(r"@\w+", '@username', text) # mengganti username
    text = re.sub(r"#\w+", '@username', text) # mengganti hastag
    # changing emoticon
    for i in range(len(emojis)):
      text = re.sub(emojis['Special Tag'][i], emojis['emot'][i], text)
    # Punctuation removal (rmoving symbols)
    text=re.sub(r"[^a-zA-Z0-9\s<>]", "", text)
    # Case folding
    text=text.lower()
    # Tokenizer
    token_text= tk.tokenize(text)
    # Stopword Removal
    token = [word for word in token_text if word not in stopword_list]
    return " ".join(token)

#TF-IDF with stemmer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import precision_recall_fscore_support
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
# With stemmer
vectorizer = TfidfVectorizer(ngram_range=(1,1), preprocessor=preprocessing_sastrawi)
print("-------with stemmer-----")
# Without stemmers
#vectorizer = TfidfVectorizer(ngram_range=(1,1), preprocessor=preprocessing_without_stemmer)
#print("-------without stemmer-----")

# Preprocess and turn into vector
X_train = vectorizer.fit_transform(train['Text Tweet'].values.tolist())


# using multinomial naive bayes
print("-------Multinomial Naive Bayes-----")
clf = MultinomialNB(alpha=0.01)
X = X_train
y = train['Sentiment'].values

clf.fit(X, y)
pred = clf.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = clf.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


#using svm
print("----------------SVM----------")
svc = SVC()
X = X_train
y = train['Sentiment'].values

svc.fit(X, y)
pred = svc.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = svc.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


-------with stemmer-----
-------Multinomial Naive Bayes-----
------train--------
              precision    recall  f1-score   support

           0       1.00      0.99      0.99       315
           1       0.99      1.00      0.99       315

    accuracy                           0.99       630
   macro avg       0.99      0.99      0.99       630
weighted avg       0.99      0.99      0.99       630

------test--------
              precision    recall  f1-score   support

           0       0.72      0.67      0.69       135
           1       0.69      0.73      0.71       135

    accuracy                           0.70       270
   macro avg       0.70      0.70      0.70       270
weighted avg       0.70      0.70      0.70       270

Confusion matrix

 [[91 44]
 [36 99]]

True Positives(TP) =  91

True Negatives(TN) =  99

False Positives(FP) =  44

False Negatives(FN) =  36
------test (in long)--------
Precision = 0.7044215626892791
recall = 0.7037037037037037
fscore = 0.703

# TF-IDF without stemmer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import precision_recall_fscore_support
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
# With stemmer
#vectorizer = TfidfVectorizer(ngram_range=(1,1), preprocessor=preprocessing_sastrawi)
# print("-------with stemmer-----")

# Without stemmers
vectorizer = TfidfVectorizer(ngram_range=(1,1), preprocessor=preprocessing_without_stemmer)
print("-------without stemmer-----")

# Preprocess and turn into vector
X_train = vectorizer.fit_transform(train['Text Tweet'].values.tolist())

# using multinomial naive bayes
print("-------Multinomial Naive Bayes-----")
clf = MultinomialNB(alpha=0.01)
X = X_train
y = train['Sentiment'].values

clf.fit(X, y)
pred = clf.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = clf.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


#using svm
print("----------------SVM----------")
svc = SVC()
X = X_train
y = train['Sentiment'].values

svc.fit(X, y)
pred = svc.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = svc.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


-------without stemmer-----
-------Multinomial Naive Bayes-----
------train--------
              precision    recall  f1-score   support

           0       1.00      0.99      1.00       315
           1       0.99      1.00      1.00       315

    accuracy                           1.00       630
   macro avg       1.00      1.00      1.00       630
weighted avg       1.00      1.00      1.00       630

------test--------
              precision    recall  f1-score   support

           0       0.71      0.67      0.69       135
           1       0.69      0.72      0.70       135

    accuracy                           0.70       270
   macro avg       0.70      0.70      0.70       270
weighted avg       0.70      0.70      0.70       270

Confusion matrix

 [[91 44]
 [38 97]]

True Positives(TP) =  91

True Negatives(TN) =  97

False Positives(FP) =  44

False Negatives(FN) =  38
------test (in long)--------
Precision = 0.6966848095002474
recall = 0.6962962962962963
fscore = 0.

# Count Vectorizer with stemmer

In [ ]:

from sklearn.metrics import precision_recall_fscore_support
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
from sklearn.feature_extraction.text import CountVectorizer
# With stemmer
vectorizer = CountVectorizer(ngram_range=(1,1), preprocessor=preprocessing_sastrawi)
print("-------with stemmer-----")
# Without stemmers
#vectorizer = CountVectorizer(ngram_range=(1,1), preprocessor=preprocessing_without_stemmer)
#print("-------without stemmer-----")

# Preprocess and turn into vector
X_train = vectorizer.fit_transform(train['Text Tweet'].values.tolist())


# using multinomial naive bayes
print("-------Multinomial Naive Bayes-----")
clf = MultinomialNB(alpha=0.01)
X = X_train
y = train['Sentiment'].values

clf.fit(X, y)
pred = clf.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = clf.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


#using svm
print("----------------SVM----------")
svc = SVC()
X = X_train
y = train['Sentiment'].values

svc.fit(X, y)
pred = svc.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = svc.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


-------with stemmer-----
-------Multinomial Naive Bayes-----
------train--------
              precision    recall  f1-score   support

           0       0.99      0.97      0.98       315
           1       0.97      0.99      0.98       315

    accuracy                           0.98       630
   macro avg       0.98      0.98      0.98       630
weighted avg       0.98      0.98      0.98       630

------test--------
              precision    recall  f1-score   support

           0       0.71      0.66      0.68       135
           1       0.68      0.73      0.70       135

    accuracy                           0.69       270
   macro avg       0.69      0.69      0.69       270
weighted avg       0.69      0.69      0.69       270

Confusion matrix

 [[89 46]
 [37 98]]

True Positives(TP) =  89

True Negatives(TN) =  98

False Positives(FP) =  46

False Negatives(FN) =  37
------test (in long)--------
Precision = 0.6934523809523809
recall = 0.6925925925925926
fscore = 0.692

# Term frequency without stemmer


In [ ]:

from sklearn.metrics import precision_recall_fscore_support
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
from sklearn.feature_extraction.text import CountVectorizer
# With stemmer
#vectorizer = CountVectorizer(ngram_range=(1,1), preprocessor=preprocessing_sastrawi)
# print("-------with stemmer-----")

# Without stemmers
vectorizer = CountVectorizer(ngram_range=(1,1), preprocessor=preprocessing_without_stemmer)
print("-------without stemmer-----")

# Preprocess and turn into vector
X_train = vectorizer.fit_transform(train['Text Tweet'].values.tolist())

# using multinomial naive bayes
print("-------Multinomial Naive Bayes-----")
clf = MultinomialNB(alpha=0.01)
X = X_train
y = train['Sentiment'].values

clf.fit(X, y)
pred = clf.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = clf.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


#using svm
print("----------------SVM----------")
svc = SVC()
X = X_train
y = train['Sentiment'].values

svc.fit(X, y)
pred = svc.predict(X)
print("------train--------")
print(classification_report(y, pred))

X_test = vectorizer.transform(test['Text Tweet'].values.tolist())
y_test = test['Sentiment'].values

pred = svc.predict(X_test)
print("------test--------")
print(classification_report(y_test, pred))


cm = confusion_matrix(y_test, pred)

print('Confusion matrix\n\n', cm)

print('\nTrue Positives(TP) = ', cm[0,0])

print('\nTrue Negatives(TN) = ', cm[1,1])

print('\nFalse Positives(FP) = ', cm[0,1])

print('\nFalse Negatives(FN) = ', cm[1,0])

print("------test (in long)--------")
precision, recall, fscore, support = precision_recall_fscore_support(y_test, pred, average='macro')
print("Precision =", precision)
print("recall =", recall)
print("fscore =", fscore)


-------without stemmer-----
-------Multinomial Naive Bayes-----
------train--------
              precision    recall  f1-score   support

           0       1.00      0.99      0.99       315
           1       0.99      1.00      0.99       315

    accuracy                           0.99       630
   macro avg       0.99      0.99      0.99       630
weighted avg       0.99      0.99      0.99       630

------test--------
              precision    recall  f1-score   support

           0       0.69      0.64      0.66       135
           1       0.66      0.70      0.68       135

    accuracy                           0.67       270
   macro avg       0.67      0.67      0.67       270
weighted avg       0.67      0.67      0.67       270

Confusion matrix

 [[87 48]
 [40 95]]

True Positives(TP) =  87

True Negatives(TN) =  95

False Positives(FP) =  48

False Negatives(FN) =  40
------test (in long)--------
Precision = 0.6746875172072022
recall = 0.674074074074074
fscore = 0.6

# using TF-IDF
## multinomial naive bayes

###with stemmer sastrawi

Precision = 0.7044215626892791\
recall = 0.7037037037037037\
fscore = 0.7034433521884782

###without stemmer

Precision = 0.6966848095002474 \
recall = 0.6962962962962963\
fscore = 0.6961462450592886

## SVM
###with stemmer sastrawi

Precision = 0.7371541501976284\
recall = 0.737037037037037\
fscore = 0.7370045684652426

###without stemmer

Precision = 0.7407935898139509\
recall = 0.7407407407407407\
fscore = 0.7407265144863915


# using Count Vectorizer
## using multinomial naive bayes
### with stemmer sastrawi
Precision = 0.6934523809523809 \
recall = 0.6925925925925926\
fscore = 0.6922506488691138
### without stemmer
Precision = 0.6746875172072022\
recall = 0.674074074074074\
fscore = 0.6737876874073261
## using SVM
### with stemmer sastrawi
Precision = 0.6821210285777215\
recall = 0.6814814814814816\
fscore = 0.6812016036026141
### without stemmer
Precision = 0.6813703537841469\
recall = 0.6777777777777778\
fscore = 0.6761741959497649